ml-kem/src/compress.rs

```diff
@@ -85,36 +91,32 @@ pub(crate) mod test {
     use super::*;
     use hybrid_array::typenum::{U1, U10, U11, U12, U4, U5, U6};
 
-    // Verify that the integer compression routine produces the same results as rounding with
-    // floats.
-    fn compression_known_answer_test<D: CompressionFactor>() {
-        let fq: f64 = FieldElement::Q as f64;
-        let f2d: f64 = 2.0_f64.powi(D::I32);
-
+    // Verify against inequality 4.7
+    fn compression_decompression_inequality<D: CompressionFactor>() {
         for x in 0..FieldElement::Q {
-            let fx = x as f64;
-            let mut x = FieldElement(x);
+            let mut y = FieldElement(x);
 
-            // Verify equivalence of compression
-            x.compress::<D>();
-            let fcx = ((f2d / fq * fx).round() as Integer) % (1 << D::USIZE);
-            assert_eq!(x.0, fcx);
+            y.compress::<D>();
+            y.decompress::<D>();
 
-            // Verify equivalence of decompression
-            x.decompress::<D>();
-            let fdx = (fq / f2d * (fcx as f64)).round() as Integer;
-            assert_eq!(x.0, fdx);
+            let lhs = (i32::from(y.0) - i32::from(x)) % (i32::from(FieldElement::Q));
+            let rhs = ((FieldElement::Q32 + (1 << (D::U32 + 1)) - 1) / (1 << (D::U32 + 1))) as i32;
```

Comment:
This is incorrect.  The accuracy threshold is supposed to be rounded.  In order to do that, you would need to make the second addend `(1 << D)` instead of `(1 << (D+1))`.

But it would be clearer to compute this with floats, and do it once at the top of the loop, since it's constant.

``` rust
// round(q / 2^(d + 1))
let accuracy = ((FieldElement::Q as f64) / ((1 << (D::U32 + 1) as f64)).round() as i32;
```

Reply:
Also, since this value is loop-invariant, I would move it to be outside the loop.